bindings: fix handling of env secrets in remote builds

ZuhairM7 opened the pull request:
Previously, using --secret=id=foo,env=BAR in remote mode would fail because the client sent the env var name to the server, which tried to resolve it locally. This patch modifies the client to resolve the environment variable locally, write it to a temp file, and send it as a file-based secret.

This aligns Podman's remote behavior with Docker's behavior and ensures secrets are correctly passed to the build container.

Test to verify fix:
`export MYSECRET='qwerty'
podman build --secret=id=mysecret,env=MYSECRET -f - . <<EOF
FROM alpine
RUN --mount=type=secret,id=mysecret cat /run/secrets/mysecret
EOF`

Before (Remote Client): The secret is empty because the server cannot find MYSECRET in its environment.
`STEP 1/2: FROM alpine
STEP 2/2: RUN --mount=type=secret,id=mysecret cat /run/secrets/mysecret
COMMIT
--> 256d6061a45d`

After (Remote Client): The client resolves MYSECRET to qwerty and sends it to the server.
`STEP 1/2: FROM alpine
STEP 2/2: RUN --mount=type=secret,id=mysecret cat /run/secrets/mysecret
qwerty
COMMIT
--> d5c40b8863ce`

Fixes #27494

<!--
Thanks for sending a pull request!

For more detailed information, please review our contributing guidelines:
https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests
-->

#### Checklist

Ensure you have completed the following checklist for your pull request to be reviewed:
<!-- Use [x] to mark as done, or click the checkbox after opening PR -->

- [x] Certify you wrote the patch or otherwise have the right to pass it on as an open-source patch by signing all
commits. (`git commit -s`). (If needed, use `git commit -s --amend`).  The author email must match
the sign-off email address. See [CONTRIBUTING.md](https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs)
for more information.
- [x] Referenced issues using `Fixes: #00000` in commit message (if applicable)
- [x] [Tests](https://github.com/containers/podman/tree/main/test#readme) have been added/updated (or no tests are needed)
- [x] [Documentation](https://github.com/containers/podman/blob/main/docs/README.md) has been updated (or no documentation changes are needed)
- [x] All commits pass `make validatepr` (format/lint checks)
- [x] [Release note](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md) entered in the section below (or `None` if no user-facing changes)

#### Does this PR introduce a user-facing change?

```release-note
Fixed an issue where `podman build --secret ... env=VAR` failed in remote mode (e.g., macOS) by correctly resolving environment variables on the client side.
```

lsm5 reviewed (commented):
@ZuhairM7 thanks for the PR. 

For fixing the `DCO` job, you'll need to signoff on your commit.
Please also check out the `Validate source code changes` job failure. That needs to get fixed to get the other jobs moving.

Can you also add a test?

mheon commented on pkg/bindings/images/build.go:
```diff
@@ -637,6 +637,20 @@ func prepareSecrets(secrets []string, contextDir string, tempManager *remote_bui
+					// read specified env into a tmp file
+					// move tmp file to tar and change secret source to relative tmp file
+					secretVal := os.Getenv(val)
+					tmpSecretFilePath, err := tempManager.CreateTempFileFromReader(contextDir, "podman-build-secret-*", strings.NewReader(secretVal))
```
Looks like this could leak - a `defer os.Remove(...)` after the error check would fix that

Honny1 reviewed (commented):
Thanks, it looks good, except for the thing the others mentioned. I fixed the release notes in the PR description.

ninja-quokka reviewed (commented):
Thanks for your first patch!

 It looks good to me and I would approve it once a test is added like @lsm5 mentioned, `test/e2e/build_test.go` looks like a good place to me.

This function could do with a refactor some time in the future.

ZuhairM7 committed:
bindings: fix handling of env secrets in remote builds

Previously, using --secret=id=foo,env=BAR in remote mode would fail because the client sent the env var name to the server, which tried to resolve it locally. This patch modifies the client to resolve the environment variable locally, write it to a temp file, and send it as a file-based secret.

Fixes #27494

Signed-off-by: ZuhairM7 <ZuhairM7>
Signed-off-by: ZuhairM7 <zuhairmerali@gmail.com>

ninja-quokka commented on test/e2e/build_test.go:
```diff
@@ -99,6 +99,15 @@ var _ = Describe("Podman build", func() {
 
+	It("podman build with a secret from env", func() {
+		os.Setenv("MYSECRET", "somesecret")
+		defer os.Unsetenv("MYSECRET")
```
Nice, we do unset some env vars on every test but not this one so it's good to unset it.

https://github.com/containers/podman/blob/4eaff6fe224bf77a4b8c732060d1da3cfec6d85d/test/e2e/common_test.go#L137-L139

ninja-quokka reviewed (approved):
/lgtm

Honny1 reviewed (commented):
LGTM. I reran the failed Packit tests.

Honny1 reviewed (approved):
LGTM

Fryguy commented:
Does this fix https://github.com/containers/podman/issues/17382 ?

Honny1 commented:
Suspect this wasn't backported. Should we backport to v5.8?

cc @mheon @TomSweeneyRedHat
